pkg/utils/landscaper/installation_simulator.go

```diff
@@ -0,0 +1,677 @@
+		tmpl, err := spiffyaml.Unmarshal(key, dataMapping.RawMessage)
+		if err != nil {
+			return fmt.Errorf("unable to parse data mapping %s for installation %s: %w", key, installationPath, err)
+		}
+		res, err := spiff.Cascade(tmpl, nil)
+		if err != nil {
+			return fmt.Errorf("unable to template data mapping %s for installation %s: %w", key, installationPath, err)
+		}
+		dataBytes, err := spiffyaml.Marshal(res)
+		if err != nil {
+			return fmt.Errorf("unable to marshal templated data mapping %s for installation %s: %w", key, installationPath, err)
+		}
+		var data interface{}
+		if err := yaml.Unmarshal(dataBytes, &data); err != nil {
+			return fmt.Errorf("unable to unmarshal templated data mapping %s for installation %s: %w", key, installationPath, err)
+		}
+		values[key] = data
+	}
+	return nil
+}
+
+// mergeMaps copies all elements of b into a.
+func mergeMaps(a, b map[string]interface{}) {
+	for key, val := range b {
+		a[key] = val
+	}
+}
+
+// executeTemplate executes a go template with the given input parameters.
+func executeTemplate(templateName, templateSource string, input map[string]interface{}) (map[string]interface{}, error) {
+	tmpl, err := gotmpl.New(templateName).Funcs(gotmpl.FuncMap(sprig.FuncMap())).Option("missingkey=zero").Parse(templateSource)
+	if err != nil {
+		parseError := gotemplate.TemplateErrorBuilder(err).WithSource(&templateSource).Build()
+		return nil, parseError
+	}
+
+	data := bytes.NewBuffer([]byte{})
+	if err := tmpl.Execute(data, input); err != nil {
+		executeError := gotemplate.TemplateErrorBuilder(err).WithSource(&templateSource).
+			WithInput(input, template.NewTemplateInputFormatter(true)).
+			Build()
+		return nil, executeError
+	}
+
+	if err := gotemplate.CreateErrorIfContainsNoValue(data.String(), templateName, input, template.NewTemplateInputFormatter(true)); err != nil {
+		return nil, err
+	}
+
+	var out map[string]interface{}
+	if err := yaml.Unmarshal(data.Bytes(), &out); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal output of export template %s: %w", templateName, err)
+	}
+
+	return out, nil
+}
+
+// convertTargetSpecToTarget converts a target spec map into a landscaper target type.
+func convertTargetSpecToTarget(name, namespace string, spec interface{}) (map[string]interface{}, error) {
+	target := lsv1alpha1.Target{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+	}
+	raw, err := yaml.Marshal(target)
+	if err != nil {
+		return nil, err
+	}
+	var unmarshalled map[string]interface{}
+	err = yaml.Unmarshal(raw, &unmarshalled)
+	if err != nil {
+		return nil, err
+	}
+	unmarshalled["spec"] = spec
+	return unmarshalled, nil
+}
+
+func encodeTemplateInput(in interface{}) (map[string]interface{}, error) {
+	raw, err := yaml.Marshal(in)
+	if err != nil {
+		return nil, err
+	}
+	var encoded map[string]interface{}
+	err = yaml.Unmarshal(raw, &encoded)
+	if err != nil {
+		return nil, err
+	}
+	return encoded, nil
+}
```

Comment:
I think such a simulator is a good idea. The only concern I have is, that even when the coding of the simulator uses the coding of the core landscaper, there is so much additional coding. Don't you have the fear that the simulator does not compute the same as the core stuff?

Reply:
I do agree with you and I share the same concern. The simulator and the landscaper core coding may diverge.
However I currently see no other way to implement such functionality when the landscaper core coding relies directly on the K8S API (which is of course understandable).

In the tests for the installation simulator I tried to use all the features that the landscaper currently provides, so that we hopefully see when things diverge.